parser/tracing.go

```diff
@@ -54,13 +56,56 @@ func (tc *tracingState) TraceDownstreamFunction() *tracingState {
 	}
 }
 
+type segmentOpts struct {
+	async  bool
+	create bool
+}
+
+// name gets the name of a segment
+func (opt *segmentOpts) name(fn *dst.FuncDecl) string {
+	name := fn.Name.Name
+	if opt.async {
+		name = fmt.Sprintf("async %s", name)
+	}
+	return name
+}
+
+// noSegment indicates that a segment should not be created
+func noSegment() segmentOpts {
+	return segmentOpts{}
+}
+
+// async segment creates a segment for an async function that begins and ends in the body of a function
+func asyncSegment() segmentOpts {
+	return segmentOpts{async: true, create: true}
+}
+
+// functionSegment creates a segment that begins and ends in the body of a function
+func functionSegment() segmentOpts {
+	return segmentOpts{create: true}
+}
+
+func prependStatementToFunctionDecl(fn *dst.FuncDecl, stmt dst.Stmt) {
+	if fn.Body == nil || fn.Body.List == nil {
+		return
+	}
+
+	fn.Body.List = slices.Insert(fn.Body.List, 0, stmt)
+}
+
+func prependStatementToFunctionLit(fn *dst.FuncLit, stmt dst.Stmt) {
+	if fn.Body == nil || fn.Body.List == nil {
+		return
+	}
+
+	fn.Body.List = slices.Insert(fn.Body.List, 0, stmt)
+}
+
 // TraceFunction adds tracing to a function. This includes error capture, and passing agent metadata to relevant functions and services.
 // Traces all called functions inside the current package as well.
 // This function returns a FuncDecl object pointer that contains the potentially modified version of the FuncDecl object, fn, passed. If
 // the bool field is true, then the function was modified, and requires a transaction most likely.
-//
-// TODO: there is a ton of complexity around tracing async statements that do not have a transaction wrapping them. This is a feature gap.
-func TraceFunction(manager *InstrumentationManager, fn *dst.FuncDecl, tracing *tracingState) (*dst.FuncDecl, bool) {
+func TraceFunction(manager *InstrumentationManager, fn *dst.FuncDecl, tracing *tracingState, segment segmentOpts) (*dst.FuncDecl, bool) {
```

Comment:
Instead of passing segment options for every trace function, would it be better to create a mapping? I think for the trace function it may be a bit more cumbersome to pass through segment options any time its called

Reply:
How would that map work? Its easy to pass it in this way because the caller of TraceFunction knowns its own scope, so it knows if a segment is needed or not

Reply:
maybe i can roll it into the tracing state object?

Reply:
I kinda prefer it as a parameter to the function. Its something temporary that needs to be set by the caller. Its the perfect parameter.

Reply:
Looking at this again, I agree with you. The reason why I was suggesting a mapping was because I was trying to imagine how the trace function would look like to support literal functions (anon functions in gin), but the solution I have in place for them doesn't depend on the trace function